sdcm/utils/remote_logger.py

```diff
@@ -302,6 +313,135 @@ def _logger_cmd(self) -> str:
+        try:
+            stream = self._k8s_core_v1_api.read_namespaced_pod_log(
+                name=self._node.name,
+                namespace=self._parent_cluster.namespace,
+                container=self._parent_cluster.container,
+                timestamps=True,
+                follow=True,
+                since_seconds=since_seconds,
+                # NOTE: need to set a timeout, because GKE's 'pod_log' API tends to hang
+                _request_timeout=self.READ_REQUEST_TIMEOUT,
+                _preload_content=False
+            )
+            self._log_reader = self._read_log_lines(stream)
+            self._reread_logs_till_last_logged_timestamp()
+        except (k8s.client.rest.ApiException, StopIteration) as exc:
+            self._log.warning(
+                "'_open_stream()': failed to open pod log stream:\n%s", exc)
+            # NOTE: following is workaround for the error 401 which may happen due to
+            #       some config data corruption during the forced socket connection failure
+            self._k8s_core_v1_api = KubernetesOps.core_v1_api(self._parent_cluster.k8s_cluster.get_api_client())
+            raise ConnectionError(str(exc)) from None
+
+    def _reread_logs_till_last_logged_timestamp(self):
+        """Upon reconnection, reread the log till last logged timestamp is found to avoid duplication"""
+        if not self._last_log_timestamp:
+            return
+        log_timestamp, line = next(self._log_reader)
+        if isoparse(log_timestamp) >= isoparse(self._last_log_timestamp):
+            self._log.debug("Current timestap is greater than last logged timestamp. No need to reread logs")
+            self._write_log_line(log_timestamp, line)
+            return
+        self._log.debug("Rereading logs till %s", self._last_log_timestamp)
+        log_timestamp = "not read yet"
+        while True:
+            for _ in range(1000):
+                log_timestamp = next(self._log_reader)[0]
+                if log_timestamp == self._last_log_timestamp:
+                    return
+            # each 1000 lines, check if we didn't pass the date, this should not happen and possibly we can drop this code after some time
+            if isoparse(log_timestamp) >= isoparse(self._last_log_timestamp):
+                TestFrameworkEvent(source="K8sClientLogger",
+                                   message=f"Failed to find last log timestamp in the log stream for {self._node.name} after reconnection."
+                                           f" Possibly some log lines are missed or duplicated. Last reread timestamp: {log_timestamp}",
+                                   severity=Severity.WARNING).publish()
+                return
+
+    def _write_log_line(self, timestamp, line):
+        self._last_read_timestamp = time.time()
+        self._last_log_timestamp = timestamp
+        self._file_object.write(line)
+
+    def _read_log_lines(self, stream) -> Generator[tuple[str, str], None, None]:
```

Comment:
From the name I expect to get a `line` on each new iteration, but on the line 409 the `chunk size` gets applied for the read operation.
Are you sure everything is ok here? Or am I missing something here?

Reply:
yes, there's `yield` 417 - so it always returns lines